dask/array/linalg.py

```diff
@@ -0,0 +1,116 @@
+from __future__ import absolute_import
+import numpy as np
+from itertools import count, product
+import dask.array as da
+import operator
+
+names = ('tsqr_%d' % i for i in count(1))
+
+
+def tsqr(data, name=None):
+    """
+    Implementation of the direct TSQR, as presented in:
+
+    A. Benson, D. Gleich, and J. Demmel.
+    Direct QR factorizations for tall-and-skinny matrices in
+    MapReduce architectures.
+    IEEE International Conference on Big Data, 2013.
+    http://arxiv.org/abs/1301.1071
+
+    :param data:
+    Shape of the blocks that will be used to compute
+    the blocked QR decomposition. We have the restrictions:
+    - blockshape[1] == data.shape[1]
+    - blockshape[0]*data.shape[1] must fit in main memory
+    :type data: dask.array.Array
+    :param name: Name of array in dask
+    :type name: basestring
+    :return: First and second tuple elements correspond to Q and R, of
+    the QR decomposition.
+    :rtype: tuple of dask.array.Array
+    """
+    if not (data.ndim == 2 and                    # Is a matrix
+            len(data.blockdims[1]) == 1):         # Only one column block
+        raise ValueError(
+            "Input must have the following properites:\n"
+            "  1. Have two dimensions\n"
+            "  2. Have only one column of blocks")
+
+    prefix = name or next(names)
+    prefix += '_'
+
+    m, n = data.shape
+    numblocks = (len(data.blockdims[0]), 1)
+
+    name_qr_st1 = prefix + 'QR_st1'
+    dsk_qr_st1 = da.core.top(np.linalg.qr, name_qr_st1, 'ij', data.name, 'ij',
+                             numblocks={data.name: numblocks})
+    # qr[0]
+    name_q_st1 = prefix + 'Q_st1'
+    dsk_q_st1 = {(name_q_st1, i, 0): (operator.getitem, (name_qr_st1, i, 0), 0)
+                 for i in xrange(numblocks[0])}
+    # qr[1]
+    name_r_st1 = prefix + 'R_st1'
+    dsk_r_st1 = {(name_r_st1, i, 0): (operator.getitem, (name_qr_st1, i, 0), 1)
+                 for i in xrange(numblocks[0])}
```

Comment:
Annoyingly dict comprehensions aren't supported in Python 2.6 (which still has enough users that we still support it).

The usual replacement is the following:

* Before:  `{k: v for ...}`
* After: `dict((k, v) for ...)`